src/main/scala/com/github/gvolpe/scalar2019/tagless/alt/dependency.scala

```diff
@@ -0,0 +1,45 @@
+package com.github.gvolpe.scalar2019.tagless.alt
+
+import cats._
+import cats.implicits._
+import scalaz.zio.{ Task, TaskR }
+
+/*
+ * Typeclass that defines the relationship between an effect that requires
+ * and environment `R` and an effect that doesn't require such environment.
+ *
+ * If we feed `R` to a given `F[R, A]` we should be able to eliminate the
+ * environment and get an `F[Any, A]`.
+ */
+abstract class MkDep[F[- _, _], R] {
```

Comment:
If you squint hard enough, it's almost like this is trying to encode half of a (less polymorphic) `MonadReader`:

```scala
trait MonadReader[F[-_, +_]] {
  val monad[R]: MonadF[R, ?]]

  def environment[R]: F[R, R]

  def provide[R, A](fra: F[R, A]): F[Any, A]
}
```

which would be the fully polymorphic version of `ApplicativeLocal`.

Reply:
I'm not sure I understand this. I'm basically trying to come up with a principled mechanism to derive `ApplicativeAsk / Local` so this takes me back to where I started :thinking: 

Anyway, as I said in the other comment I'll be thinking about this a bit more and see what comes up.